### DISCUSSION TITLE
How to get full resolution result of geo_velocity.h5?

### DISCUSSION THREAD

**Discussion by gjustin40** [insarlab/MintPy#1106] [Q&A]
The starting point for this question is here. **[Miaplpy Example](https://github.com/insarlab/MiaplPy/issues/70)**

I am trying to Visualize the result of land displacement with tif data, How can i get the resolution that I prefer?

From above link, I found it that the key of resolution of result is `mintpy.geocode.laloStep`
I tried several cases with command `smallbaselineApp.py factory.txt --dir ./miaplpy2/network_single_reference/ --start geocode` by:
```
mintpy.geocode.laloStep     = 0.000092593,0.000092593 (which is 10m)
mintpy.geocode.laloStep     = 0.000555556,0.000555556 (which is 50m)
mintpy.geocode.laloStep     = 0.00092593,0.00092593 (which is 100m)
```
and visualized `geo_velocity.h5` with `save_kmz.py` and upload to google earth. the results are:
![image](https://github.com/insarlab/MintPy/assets/25999141/6619332b-ebb0-46ec-9aaa-5d9a5a8a577a)
(100m, 50m, 10m)

But when I use `save_gdal.py` function to result of each options, the resolution of tif data are same on QGIS
Furthermore, the resolution is different from what I expected.
![image](https://github.com/insarlab/MintPy/assets/25999141/8768d248-aa9e-4241-8501-b8a475c3ee78)

**I want to generate the tif data for visualize the resolut of land displacement with the resolution the i expected, 
How to do that?**

Thanks

**Comment by yunjunz** [comment]
Good question. By default, mintpy/geocode.py uses [this approach to determine the output pixel size](https://github.com/insarlab/MintPy/blob/2f7797dd3fc818137f5cc7a0e031f283171c5b97/src/mintpy/objects/resample.py#L293-L299), which uses the following criteria:
1. the output pixel should have the same area as the input pixel
2. the output pixel lat/lon ratio matches the height/width of the bounding box of the radar pixel on the ground, as shown here (https://math.stackexchange.com/questions/4001034/calculate-the-dimensions-of-a-rotated-rectangle-inside-a-bounding-box).

Therefore, for a full-resolution product in geo-coordiantes, I think the default output grid (without manually setting the `mintpy.geocode.laloStep` option) is your best choice. And note that, the output grid in geo-coordinates will not be exactly 5 x 20 m, but should be close to that. Another point is that mintpy uses the nearest resampling only during the geocoding, thus, setting a `mintpy.geocode.laloStep` value smaller than the actual pixel size, as in the case of 10 m, will only duplicate some values in nearby pixels.

Alternatively, if you only need to show the data in KML, using the radar-coded velocity file, as in `save_kmz.py velocity.h5` would avoid the resampling procedure, thus, no issue of choosing an output pixel size.

**Reply by gjustin40** [reply]
Thanks for your informations.

I want to generate TIFF data with a resolution as close as possible to `20x5m`, although it doesn't need to be an exact match. 
However, as you can see in the image above, around `24x40` seems to have too much difference.

What should I do to create a TIFF as close as possible using `save_gdal.py`?

Thanks

**Reply by yunjunz** [reply]
Make sure your `AZIMUTH/RANGE_PIXEL_SIZE` metadata is correct in your `velocith.h5` file (they should be close to 2.3 and 14.1), by running `info.py velocity.h5` to check.

1. Unset `mintpy.geocode.laloStep` option in `smallbaselinApp.cfg`; run `smalllbaselineApp.py` or `geocode.py velocity.h5 -o geo/geo_velocity.h5`
2. Run `save_gdal.py geo/geo_velocity.h5`

**Comment by falkamelung** [comment]
We just use the radar-coded files and then use the lat/long information for each pixel to plot them on a map (aka geocoding).  We have code for plotting but it is not in a good shape.  I can share though if you are interested.

**Reply by gjustin40** [reply]
Thanks for your support. @falkamelung 

If possible, I would like to receive the code for this.
Thanks

**Comment by falkamelung** [comment]
Here the two scripts. The first plots velocity or DemError on a map. The second also plots the time series. But they are messy and the python needs re-factoring. You are of course welcome to do this. The second one will be renamed to view_full_res_timeseries.py or so.

https://github.com/geodesymiami/rsmas_insar/blob/master/minsar/view_scatterplot.py
https://github.com/geodesymiami/rsmas_insar/blob/master/minsar/view_ts.py
